src/animation.cpp

```diff
@@ -0,0 +1,429 @@
+#include "game.h"
+#include "debug.h"
+#include "cata_tiles.h" // all animation functions will be pushed out to a cata_tiles function in some manner
+
+extern cata_tiles *tilecontext; // obtained from sdltiles.cpp
+extern void try_update();
+
+// see game.cpp
+bool is_valid_in_w_terrain(int x, int y);
+
+namespace {
+void draw_animation_delay(long const scale = 1)
+{
+    auto const delay = static_cast<long>(OPTIONS["ANIMATION_DELAY"]) * scale * 1000000l;
+
+    timespec const ts = {0, delay};
+    if (ts.tv_nsec > 0) {
+        nanosleep(&ts, nullptr);
+    }
+}
+
+void draw_explosion_curses(WINDOW *const w, int const x, int const y,
+    int const r, nc_color const col)
+{
+    if (r == 0) { // TODO why not always print '*'?
+        mvwputch(w, x, y, col, '*');
+    }
+
+    for (int i = 1; i <= r; ++i) {
+        mvwputch(w, y - i, x - i, col, '/');  // corner: top left
+        mvwputch(w, y - i, x + i, col, '\\'); // corner: top right
+        mvwputch(w, y + i, x - i, col, '\\'); // corner: bottom left
+        mvwputch(w, y + i, x + i, col, '/');  // corner: bottom right
+        for (int j = 1 - i; j < 0 + i; j++) {
+            mvwputch(w, y - i, x + j, col, '-'); // edge: top
+            mvwputch(w, y + i, x + j, col, '-'); // edge: bottom
+            mvwputch(w, y + j, x - i, col, '|'); // edge: left
+            mvwputch(w, y + j, x + i, col, '|'); // edge: right
+        }
+    }
+    
+    wrefresh(w);
+    draw_animation_delay(EXPLOSION_MULTIPLIER);
+}
+} // namespace
+
+#if !defined(SDLTILES)
+void game::draw_explosion(int const x, int const y, int const r, nc_color const col)
+{
+    const int ypos = POSY + (y - (u.posy() + u.view_offset_y));
+    const int xpos = POSX + (x - (u.posx() + u.view_offset_x));
+    draw_explosion_curses(w_terrain, xpos, ypos, r, col);
+}
+#else
+void game::draw_explosion(int const x, int const y, int const r, nc_color const col)
+{
+    // added offset values to keep from calculating the same value over and over again.
+    const int ypos = POSY + (y - (u.posy() + u.view_offset_y));
+    const int xpos = POSX + (x - (u.posx() + u.view_offset_x));
+
+    if (!use_tiles) {
+        draw_explosion_curses(w_terrain, xpos, ypos, r, col);
+        return;
+    }
+
+    for (int i = 1; i <= r; i++) {
+        tilecontext->init_explosion(x, y, i); // TODO not xpos ypos?
```

Comment:
> TODO not xpos ypos?

Because `cata_tiles::draw_explosion_frame` does this calculation on its own. Specifically it's done as the last step in `cata_tiles::draw_from_id_string`. This can not be avoided as all tile drawing goes through that function, `cata_tiles` would have to revers that calculation when it initiates the tile drawing.

It also seems you can move the calculation of `xpos` and `ypos` into `draw_explosion_curses`.